rpc: treat univalue type check error as RPC_TYPE_ERROR, not RPC_MISC_ERROR

furszy opened the pull request:
Same rationale as #26039, tackling another angle of the problem.

#### Context
We have the same univalue type error checking code spread/duplicated few times:
`RPCTypeCheckObj`, `RPCTypeCheckArgument`, `UniValue::checkType`.

In the first two functions, we are properly returning an `RPC_TYPE_ERROR` while in `UniValue::checkType`
we are throwing an `std::runtime_error` which is caught by the RPC server request handler, who invalidly
treats it as `RPC_MISC_ERROR` (which is a generic error return code that provides no information to the user).

#### Proposed Changes

Throw a custom exception from `Univalue::checkType` (instead of a plain
`std::runtime_error`) and catch it on the RPC server request handler.

So we properly return `RPC_TYPE_ERROR` (-3) on every arg type error and
not the general `RPC_MISC_ERROR` (-1).

This will allow us to remove all the `RPCTypeCheckArgument` calls. As them are redundant since #25629.

furszy committed:
rpc: remove unneeded RPCTypeCheckArgument checks

No-behavior change.

Since #25629, we check the univalue type internally.

fanquake commented:
In https://cirrus-ci.com/task/5952817063002112 (mempool_accept.py):
```bash
 test  2022-07-29T13:45:02.711000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 135, in try_rpc
                                       fun(*args, **kwds)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/mempool_accept.py", line 68, in <lambda>
                                       assert_raises_rpc_error(-3, 'JSON value of type string is not of expected type array', lambda: node.testmempoolaccept(rawtxs='ff00baar'))
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/coverage.py", line 49, in __call__
                                       return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 144, in __call__
                                       raise JSONRPCException(response['error'], status)
                                   test_framework.authproxy.JSONRPCException: Expected type array, got string (-3)
                                   During handling of the above exception, another exception occurred:
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 133, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/mempool_accept.py", line 68, in run_test
                                       assert_raises_rpc_error(-3, 'JSON value of type string is not of expected type array', lambda: node.testmempoolaccept(rawtxs='ff00baar'))
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 126, in assert_raises_rpc_error
                                       assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 143, in try_rpc
                                       message, e.error['message']))
                                   AssertionError: Expected substring not found in error message:
                                   substring: 'JSON value of type string is not of expected type array'
                                   error message: 'Expected type array, got string'.
```

furszy commented:
thx @fanquake, pushed 👍🏼.

jonatack reviewed (commented):
Would it make sense to first adapt the UniValue error messages so that these remain the same for client software? Otherwise this seems like it would be needlessly changing user space.

furszy commented:
> Would it make sense to first adapt the UniValue error messages so that these remain the same for client software? Otherwise this seems like it would be needlessly changing user space.

good point @jonatack. Will move into that direction.

still, the new error introduced in 25629 is much more descriptive (and helpful) than the current one. Might be good to deprecate the old one and move to the new format.

fanquake commented:
> good point @jonatack. Will move into that direction.
> still, the new error introduced in https://github.com/bitcoin/bitcoin/pull/25629 is much more descriptive, and helpful for the user, than the current one. Might be good to deprecate the old one and move to the new format.

I really don't think we need to revert these changes (£25629), or have any sort of "deprecation" for the older messages.

MarcoFalke commented on src/univalue/lib/univalue.cpp:
```diff
@@ -217,8 +217,7 @@ const UniValue& UniValue::operator[](size_t index) const
     if (typ != expected) {
-        throw std::runtime_error{"JSON value of type " + std::string{uvTypeName(typ)} + " is not of expected type " +
-                                 std::string{uvTypeName(expected)}};
+        throw std::runtime_error{"Expected type " + std::string{uvTypeName(expected)} +  ", got " + std::string{uvTypeName(typ)}};
```
ACK, previously this was `"JSON value is not an {} as expected"`. Then, changed in fae5ce8795080018875227aee8613677f92e99ce.

I think I wanted to use your wording, so mind to split this up into a separate pull?

furszy replied:
will just drop the revert (rationale: https://github.com/bitcoin/bitcoin/pull/25737#issuecomment-1200218274).

MarcoFalke replied:
Sure, no strong opinion. Just to clarify, you haven't "dropped the revert": It is not a revert, but a change to the string. Also, the string is still changed in this pull request.

furszy replied:
> Just to clarify, you haven't "dropped the revert": It is not a revert, but a change to the string. 

yeah ok. I thought that the string before fae5ce87 was the "Expected type {expected], got {type}" thus why was thinking on a revert but seems that not. We actually have two different errors for the same problematic.
Depends on the RPC command, we either throw a "JSON value of type {type} is not of expected type {expected}" or the "Expected type {expected], got {type}".

> Also, the string is still changed in this pull request.

Yeah, that is intended.

Unless someone has a strong opinion about it, will actually move the sources to only use the "JSON value of type {type} is not of expected type {expected}" error description everywhere and drop the other, less informative, one.

I think that It doesn't make sense to build any client based on a json parsing error (if the command is expecting an int and you provide a string, the command will fail anyway..), so no one should be affected by this change.

furszy commented:
> > good point @jonatack. Will move into that direction.
> > still, the new error introduced in #25629 is much more descriptive, and helpful for the user, than the current one. Might be good to deprecate the old one and move to the new format.
> 
> I really don't think we need to revert these changes (£25629), or have any sort of "deprecation" for the older messages.

Thinking about this for a second time, and.. agree.  It doesn't make sense to build any client based on a json parsing error.. if the RPC command is expecting an int and receives a string, it will always fail and never execute the command. So the new error description shouldn't affect anyone. Will just drop the first commit.

jonatack commented:
If we really want to change user space, would it make sense to opt for the most frequently used of the error message formats currently, and the shortest and most straightforward one?

jonatack commented:
(See https://github.com/bitcoin/bitcoin/pull/24944#discussion_r857437282 for a list of RPCs.)

furszy commented:
> If we really want to change user space, would it make sense to opt for the most frequently used of the error message formats currently, and the shortest and most straightforward one?

Right now, we are throwing two different errors for the same problematic.

1) "Expected type {expected], got {type}" 
2) "JSON value of type {type} is not of expected type {expected}"

This PR aims to drop the first one as it is less informative than the second one.

aureleoules commented on src/rpc/util.cpp:
```diff
@@ -41,19 +41,15 @@ void RPCTypeCheck(const UniValue& params,
 
         const UniValue& v = params[i];
         if (!(fAllowNull && v.isNull())) {
-            RPCTypeCheckArgument(v, t);
+            if (!t.typeAny && v.type() != t.type) {
+                throw JSONRPCError(RPC_TYPE_ERROR,
+                                   strprintf("JSON value of type %s is not of expected type %s", uvTypeName(v.type()), uvTypeName(t.type)));
+            }
```
maybe merge if statements?

furszy replied:
I'm unsure about chaining all those statements together. It's hard to read already.

What about splitting the statements?
e.g.
```c++
if (fAllowNull && v.isNull()) continue;
if (!t.typeAny && v.type() != t.type) {
    throw JSONRPCError(RPC_TYPE_ERROR,
          strprintf("JSON value of type %s is not of expected type %s", uvTypeName(v.type()), uvTypeName(t.type)));
}
```

aureleoules replied:
looks good :+1:

aureleoules reviewed (commented):
ACK d54f65c5d10983b97c8d33273a62b29b70b3853a.
Verified that `RPCTypeCheckArgument` calls are redundant.
+1 for unifying error messages.

jarolrod commented:
concept ack

MarcoFalke commented on src/rpc/mempool.cpp:
```diff
@@ -605,7 +605,6 @@ static RPCHelpMan gettxspendingprevout()
         },
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
         {
-            RPCTypeCheckArgument(request.params[0], UniValue::VARR);
```
why is this redundant?

furszy replied:
oops, great catch. There is another one in coins.cpp. Pushing..

MarcoFalke replied:
Now you changed the error code from -3 to -1, effectively the opposite of what #26039 is doing

furszy replied:
true.. general exceptions are -1, while this is a type error.

We could throw a custom exception on the UniValue type check to catch all of them as type errors.

But.. I'm more inclined to close this PR and directly go ahead with moving the type checks responsibility to the request handler (which seems that is what you are doing in #26039).
Haven't checked if there is any restriction there but we do have each arg type on the help info object, so.. we could use that to verify every arg type.

MarcoFalke replied:
now you are changing the behavior in commit 8bdd192d4f897cc872f33cd6d15c36449dc345e5 and then reverting the behavior change in the next commit?

maflcko reviewed (changes requested):
.

furszy commented:
pushed per @MarcoFalke feedback. Thanks 👌🏼.

MarcoFalke commented on src/rpc/util.cpp:
```diff
@@ -40,20 +40,15 @@ void RPCTypeCheck(const UniValue& params,
+        if (fAllowNull && v.isNull()) continue;
+        if (!t.typeAny && v.type() != t.type) {
+            throw JSONRPCError(RPC_TYPE_ERROR,
+                               strprintf("JSON value of type %s is not of expected type %s", uvTypeName(v.type()), uvTypeName(t.type)));
```
Would be nice to split this behaviour change of unifying the string into a separate pull request from the refactoring changes

furszy replied:
ok, one sec.

furszy commented:
Based on Marko feedback, have removed the string unification from this PR.
Will push it in a subsequent work. Ready to go.

MarcoFalke commented:
> have removed the string unification from this PR. Will push it in a subsequent work.

Are you still working on the string unification?

fanquake commented:
Might be worth closing / combining this into #26039? That PR is a more comprehensive version of the same kind of change.

furszy commented:
> Might be worth closing / combining this into #26039? That PR is a more comprehensive version of the same kind of change.

yeah, I wrote the same few days ago to MarkoFalke here https://github.com/bitcoin/bitcoin/pull/25737#discussion_r967650202.

Still, another proposal would be to do something like this: https://github.com/furszy/bitcoin/commit/e6720177b768d79b7f59f5dc6f77d1527b23fc05
Essentially, throw a custom exception for the type error instead of a bare `runtime_error`.

MarcoFalke commented:
> throw a custom exception

Yeah, this shouldn't hurt. Though I think there is still a benefit to checking the types in the rpc help manager:

* In the future more complicated types can be checked, for example hex strings or amounts
* Checking the types in the beginning may avoid a costly abort (when calculations have been done) or brittle abort (when resources need to be freed) when the type error is found later

furszy commented:
> * Checking the types in the beginning may avoid a costly abort (when calculations have been done) or brittle abort (when resources need to be freed) when the type error is found later

Absolutely. The RPC framework should at least verify that the univalue types are the expected ones.

> In the future more complicated types can be checked, for example hex strings or amounts

The only nano-concern that have about this is that we could end up downgrading performance by parsing the arguments twice (first at the framework level, then at the RPC command). So, we would need to store the parsed values somewhere in the request context.

-------

So, options for this PR:
(1) I could close it in favor of #26039 or.. (2) could push the univalue type error custom exception (this would make #26039 smaller).

Agree on going with (2) or do you prefer (1)? (I'm fine either way)

MarcoFalke commented:
I think both can be done independently

furszy committed:
rpc: treat univalue type check error as RPC_TYPE_ERROR, not RPC_MISC_ERROR

By throwing a custom exception from `Univalue::checkType` (instead of a plain
std::runtime_error) and catching it on the RPC server request handler.

So we properly return RPC_TYPE_ERROR (-3) on arg type errors and
not the general RPC_MISC_ERROR (-1).

MarcoFalke commented on src/univalue/include/univalue.h:
```diff
@@ -19,6 +19,15 @@ class UniValue {
 public:
     enum VType { VNULL, VOBJ, VARR, VSTR, VNUM, VBOOL, };
 
+    class type_error : public std::exception
+    {
+    private:
+        const std::string msg;
+    public:
+        explicit type_error(const std::string& _err) : msg(_err) {}
+        const char* what() const noexcept override { return msg.c_str(); }
+    };
+
```
```suggestion
    class type_error : : public std::runtime_error
{
    using std::runtime_error::runtime_error;
};
```

furszy replied:
k

MarcoFalke commented on src/rpc/util.cpp:
```diff
@@ -40,21 +40,15 @@ void RPCTypeCheck(const UniValue& params,
     }
 }
 
-void RPCTypeCheckArgument(const UniValue& value, const UniValueType& typeExpected)
```
not really a fan of removing this, given that this is needed for https://github.com/bitcoin/bitcoin/pull/26039

furszy replied:
Left you a comment there, https://github.com/bitcoin/bitcoin/pull/26039#discussion_r971027963.
I think that we shouldn't continue having the type error string defined in two different places.

MarcoFalke replied:
I am removing `RPCTypeCheck` and you are adding the type error string to `RPCTypeCheck`. So I don't see how the changes here require removing it or help in any way to reducing it to one place.

MarcoFalke replied:
Also, unrelated: `RPCTypeCheckObj` has a third and different type error string?

furszy replied:
> I am removing `RPCTypeCheck` and you are adding the type error string to `RPCTypeCheck`. So I don't see how the changes here require removing it or help in any way to reducing it to one place.

Yeah, it's not really needed here. I was pointing out that, as you are removing `RPCTypeCheck`, and I'm removing `RPCTypeCheckArgument` here, we could also reduce the error string duplication there as well.

But.. I could also implement the changes here too..
So yeah, probably better to leave it up for another PR and done.

> Also, unrelated: RPCTypeCheckObj has a third and different type error string?

Yeah, I mentioned it in one of the first comments of this PR https://github.com/bitcoin/bitcoin/pull/25737#issuecomment-1200425927 and then forgot about it.
Could fix it on the follow-up `Univalue::checkType` unification PR.

maflcko reviewed (commented):
left two nits

MarcoFalke commented:
Also, the tests fail

MarcoFalke commented:
It would be good to explain the behavior change in the title, not summarize the detailed code changes. If someone was interested in the detailed code changes they could look at them, or you could summarize them in the commit/pull description.

Also, it would help reviewers if you explained for each commit whether it is a refactor or behavior change. This helps to document whether a change was intentional or not.

furszy commented:
> It would be good to explain the behavior change in the title, not summarize the detailed code changes. If someone was interested in the detailed code changes they could look at them, or you could summarize them in the commit/pull description.

Sure, the lack of scope in the PR description is mainly because this started as a simple redundant code removal (which didn't require much text) and then we realized that for making that cleanup, needed to properly return `RPC_TYPE_ERROR` instead of `RPC_MISC_ERROR` in the inner univalue type check.

> Also, it would help reviewers if you explained for each commit whether it is a refactor or behavior change. This helps to document whether a change was intentional or not.

Sure.

furszy commented:
Updated per feedback.
